File: apps/web/types/dance.ts
export type DanceDifficulty = "easy" | "medium" | "hard";
export type DanceStyle = "fun" | "cool" | "cute" | "high_energy";

export type DanceMove = {
  id: string;
  moveId: string;
  label: string;
  startBeat: number;
  endBeat: number;
  durationBeats: number;
  animationClip: string;
};

export type SongAnalysis = {
  songId: string;
  bpm: number;
  durationSeconds: number;
  beats: number[];
};

export type DancePlan = {
  id: string;
  songId: string;
  bpm: number;
  style: DanceStyle;
  difficulty: DanceDifficulty;
  totalBeats: number;
  moves: DanceMove[];
};

File: apps/api/models/dance.py
from typing import List, Literal

from pydantic import BaseModel

DanceDifficulty = Literal["easy", "medium", "hard"]
DanceStyle = Literal["fun", "cool", "cute", "high_energy"]


class GenerateDanceRequest(BaseModel):
    songId: str
    bpm: int = 120
    style: DanceStyle = "fun"
    difficulty: DanceDifficulty = "easy"
    totalBeats: int = 8


class DanceMove(BaseModel):
    id: str
    moveId: str
    label: str
    startBeat: int
    endBeat: int
    durationBeats: int
    animationClip: str


class DancePlan(BaseModel):
    id: str
    songId: str
    bpm: int
    style: DanceStyle
    difficulty: DanceDifficulty
    totalBeats: int
    moves: List[DanceMove]
